# 숫자형
- 숫자형은 숫자 형태로 이루어진 자료형

| 데이터 유형 |          |
|-------------|----------|
| 정수        | `x = 10`, `y = -15`   |
| 실수        | `a = 3.14`, `b = -0.005` |
| 8진수       | `oct_num = 0o10`, `oct_num2 = 0o777` |
| 16진수      | `hex_num = 0x1F`, `hex_num2 = 0xABCD` |

- 8진수 : 
    - 8진수는 0부터 7까지의 숫자로 표현되는 수 체계이며, 컴퓨터 과학에서 주로 사용됩니다. 
    - 10진수에서 8진수로 변환하려면 8로 나누어가며 나머지를 구하면 됩니다. 
    - 8진수는 주로 파일 권한 설정과 같은 시스템 관련 작업에서 사용됩니다.
- 16진수 : 
    - 16진수는 0부터 9까지의 숫자와 A부터 F까지의 문자를 사용하여 나타내는 수 체계입니다. 
    - 이진수와 함께 주로 컴퓨터 과학 및 프로그래밍에서 사용되며, 16진수는 이진수로 변환하기가 쉽고 간단하여 메모리 주소, RGB 색상 코드 및 문자 코드와 같은 다양한 데이터를 표현하는 데에 유용하게 사용됩니다. 
    - 16진수는 10진수에서 16진수로 변환하려면 16으로 나누어가며 나머지를 구하면 됩니다.    

## 숫자형은 어떻게 만들고 사용할까?

**정수형**
- 정수형(integer)란 정수를 뜻하는 자료형

In [1]:
a = 123

In [2]:
a = -178

In [3]:
a = 0

**실수형**
- 실수형(floating point) 는 소수점이 포함된 숫자
- 이 표현 방식은 실수를 가수와 지수의 형태로 표현합니다. 
- 여기서 "가수"는 소수점 이하의 숫자를 표현하고, "지수"는 소수점의 위치를 나타냅니다.
- 예를 들어, 3.14를 부동 소수점으로 표현하면 가수는 3.14의 가수 부분이 되고, 지수는 소수점의 위치를 조절하여 정확한 값을 표현

In [6]:
a = 0.12

In [7]:
a = -3.45

- 컴퓨터 지수 표현 방식

In [10]:
a =  # 1.2*10^-1

In [11]:
a =  # -3.45*10^0

**8진수와 16진수**

- 8진수를 만들기 위해서는 0o 또는 0O(숫자 0과 소문자 o 또는 대문자 O)로 시작

In [12]:
a = 
a

127

- 16진수를 만들기 위해서는 0x로 시작

In [13]:
a = 
b = 
print(a, b)

2303 2748


## 숫자형을 활용하기 위한 연산자

**사칙 연산**

In [14]:
a = 3
b = 4


7

-1

12

0.75

**x의 y제곱을 나타내는 ** 연산자**

In [4]:
a = 3
b = 4


81

**문제**
- 방금 배운 연산자를 활용 $10\times18^2+2 \times 11$를 계산해 보자.

**나눗셈 후 나머지를 리턴하는 % 연산자**
- % 연산자는 나눗셈의 나머지 값을 리턴하는 연산
- 7을 3으로 나누면 1

1

- 3을 7로 나누면 3

3

**나눗셈 후 몫을 리턴하는 // 연산**
- 7/4

1.75

- 7를 4로 나눈 몫

1

**문제**
- 14를 3으로 나누었을 때 몫과 나머지를 구하라

# 문자열 자료형
- [소스](https://wikidocs.net/13)
- 문자열(string)이란 문자, 단어 등으로 구성된 문자들의 집합

    - "Life is too short. You need Python"
    - "a"
    - "123"

- 모든 예문은 큰따움표로 둘러싸여 있다.
- "123"은 숫자인데 왜 문자열일까?

## 문자열을 어떻게 만들고 사용할 까?
- 다음과 같이 4가지가 있다

**큰따움표로 양쪽 둘러싸기**

'Hello world'

**작은따움표로 양쪽 둘러싸기**

'Python is fun'

**큰따움표 3개를 연속으로 써서 양쪽 둘러싸기**

'Life is too short. You need python.'

**작은따움표 3개를 연속으로 써서 양쪽 둘러싸기**

'Life is too short. You need python.'

## 문자열 안에 작은따움표나 큰따움표를 포함시키고 싶을 때
**1.문자열에 작은따움표 포함하기**

- Python's favorite food is perl

"Python's favorite food is perl"

SyntaxError: unterminated string literal (detected at line 1) (1974562179.py, line 1)

**2.문자열에 큰따움표 포함하기**
- "Python is very easy." he says.

'"Python is very easy." he says.'

**3.역슬래시를 사용 작은따움표와 큰따움표를 문자열에 포함시키기**
- 역슬래쉬 뒤의 "나 ' 는 문자열을 둘러싸는 기호가 아니라 문자 그 자체가 된다.
    - 역슬래쉬는 많은 프로그래밍 언어에서 이스케이프(escape) 문자입니다. 
    - 이스케이프 문자는 특수한 기능을 가진 문자를 일반 문자로 해석하도록 지시합니다. 
    - 즉, 그 다음에 오는 문자를 특별한 의미를 갖지 않고 그대로 문자열로 처리하도록 합니다.

"Python's favorite food is perl"

'"Pythons is very easy." he says.'

## 여러 줄인 문자열을 변수에 대입하고 싶을 때

- 다음과 같이 여러줄을 변수에 대입하고자 한다면 어떻게 해야 할까.

Life is too short
You need python


**1.줄을 바꾸기 위한 이스케이프 코드 \n 삽입**

Life is too short
You need python


**2.연속된 작은따움표 3개 또는 큰따움표ㅕ3개 사용하기**

Life is too short
You need python


Life is too short
You need python


**이스케이프 코드**

- 이스케이프 코드란 프로그래밍 할 때 사용할 수 있도록 미리 정해 놓은 '문자 조합'
- 주로 출력물을 보기 좋게 정렬하는 용도

| 코드  | 설명                                 |
|:------|:-------------------------------------|
| \n    | 문자열 안에서 줄을 바꿀 때 사용      |
| \t    | 문자열 사이에 탭 간격을 줄 때 사용  |
| \\    | \를 그대로 표현할 때 사용            |
| \'    | 작은따옴표(')를 그대로 표현할 때 사용 |
| \"    | 큰따옴표(")를 그대로 표현할 때 사용 |
| \r    | 캐리지 리턴(줄 바꿈 문자)           |
| \f    | 폼 피드(줄 바꿈 문자)               |
| \a    | 벨 소리(출력할 때 PC 스피커에서 '삑' 소리가 난다) |
| \b    | 백 스페이스                         |
| \000  | 널 문자                             |

- \n, \t, \\, \', \" 는 자주 사용되니 기억해 두자

## 문자열 연산하기
- 문자열을 더하거나 곱해보자

**문자열 더해서 연결하기**

'Python is fun!'

**문자열 곱하기**

'PythonPython'

**문자열 곱하기 응용**

- 다음을 실행해 보자

```python

print("=" * 50)
print("My Program")
print("=" * 50)

```

**문자열 길이 구하기**
- 문자열의 길이는 다음과 같이 len 함수를 사용하면 구할 수 있다. 
- len 함수는 print 함수처럼 파이썬의 기본 내장 함수로, 별다른 설정 없이 바로 사용할 수 있다.

> 문자열의 길이에는 공백 문자도 포함된다.

17

**문제**
- "You need python" 이라는 문장을 문자열로 만들고, 길이를 구해보자.

## 문자열 인덱싱과 슬라이싱
- 인덱싱(indexing)이란 무엇인가를 ‘가리킨다’, 슬라이싱(slicing)은 무엇인가를 ‘잘라 낸다’라는 의미이다. 
- 이런 의미를 생각하면서 다음 내용을 살펴보자.

**문자열 인덱싱**

- 위 코드에서 변수 a에 저장한 문자열의 각 문자마다 번호를 매겨 보면 다음과 같다.

<img src="https://wikidocs.net/images/page/13/02_2_indexing.png">

- "Life is too short, You need Python" 문자열에서 L은 첫 번째 자리를 뜻하는 숫자 0, i는 1 이런 식으로 계속 번호를 붙인 것이다. 
- 즉, 중간에 있는 short의 s는 12가 된다.

'e'

- a[3]이 뜻하는 것은 a라는 문자열의 네 번째 문자 e를 말한다. 
- 프로그래밍을 처음 접하는 독자라면 a[3]에서 숫자 3이 왜 네 번째 문자를 뜻하는지 의아할 수도 있다. 
- 사실 이 부분이 헷갈릴 수 있는 부분인데, 다음과 같이 생각하면 쉽게 알 수 있을 것이다.

**"파이썬은 0부터 숫자를 센다."**

- 따라서 파이썬은 위 문자열을 다음과 같이 바라보고 있다.

> a[0]:'L', a[1]:'i', a[2]:'f', a[3]:'e', a[4]:' ', ...

- 0부터 숫자를 센다는 것이 처음에는 익숙하지 않겠지만, 계속 사용하다 보면 자연스러워질 것이다. 
- 위 예에서 볼 수 있듯이 a[번호]는 문자열 안의 특정한 값을 뽑아 내는 역할을 한다. 이러한 작업을 **‘인덱싱’** 이라고 한다.

**문자열 인덱싱 활용하기**

'L'

's'

'n'

- a[-1]이 뜻하는 것은 뭘까?
    - 문자열을 뒤에서부터 읽기 위해 -(빼기) 기호를 붙인 것
    - a[-1]은 뒤에서부터 세어 첫 번째가 되는 문자를 말한다. 
    - a의 값은 "Life is too short, You need Python" 문자열이므로 뒤에서부터 첫 번째 문자는 가장 마지막 문자 'n'이다.
    - 뒤에서부터 첫 번째 문자를 표시할 때도 0부터 세어 ‘a[-0]이라고 해야 하지 않을까?’라는 의문이 들 수도 있겠지만, 잘 생각해 보자. 
    - 0과 -0은 똑같은 것이기 때문에 a[-0]은 a[0]과 똑같은 값을 보여 준다.

'L'

'o'

'y'

**문자열 슬라이싱**
- 그렇다면 "Life is too short, You need Python" 문자열에서 단순히 한 문자만을 뽑아 내는 것이 아니라 ‘Life’ 또는 ‘You’와 같은 단어를 뽑아 내는 방법은 없을까?

'Life'

- 위와 같이 단순하게 접근할 수도 있지만 파이썬에서는 더 좋은 방법을 제공한다. 바로 슬라이싱(slicing) 기법이다. 
- 위 예는 슬라이싱 기법으로 다음과 같이 간단하게 처리할 수 있다.

> 인덱싱 기법과 슬라이싱 기법은 뒤에서 배울 자료형인 리스트나 튜플에서도 사용할 수 있다.

'Life'

- a[0:4]는 a 문자열, 즉 "Life is too short, You need Python" 문자열에서 자리 번호 0부터 4까지의 문자를 뽑아 낸다는 뜻이다.
- 하지만 ‘a[0]은 L, a[1]은 i, a[2]는 f, a[3]은 e이므로 a[0:3]으로도 Life라는 단어를 뽑아 낼 수 있지 않을까?’라는 의문이 생길 것이다. 
- 다음 예로 확인해 보자.

'Lif'

- 이렇게 되는 이유는 슬라이싱 기법으로 a[시작_번호:끝_번호]를 지정할 때 끝 번호에 해당하는 문자는 포함하지 않기 때문이다. 
- 즉, a[0:3]을 수식으로 나타내면 다음과 같다.

$0 <= a < 3$

- 이 수식을 만족하는 것은 a[0], a[1], a[2]이다. 
- 따라서 a[0:3]은 'Lif', a[0:4]는 'Life'가 되는 것이다. 
- 이 부분이 문자열 연산에서 가장 혼동하기 쉬운 부분이므로 문제를 많이 풀어 보면서 몸에 익히기 바란다.

**문자열을 슬라이싱하는 방법**

'Life '

- 위 예는 a[0] + a[1] + a[2] + a[3] + a[4]와 동일하다. 
- a[4]는 공백 문자이기 때문에 'Life'가 아닌 'Life '가 출력된다. 
- 공백 문자 역시 L, i, f, e 와 같은 문자와 동일하게 취급되는 것을 잊지 말자. 
- 'Life'와 'Life '는 완전히 다른 문자열이다.

- 슬라이싱할 때 항상 시작 번호가 0일 필요는 없다.

'Li'

'is'

'short'

- a[시작_번호:끝_번호]에서 끝 번호 부분을 생략하면 시작 번호부터 그 문자열의 끝까지 뽑아 낸다.

'You need Python'

- a[시작_번호:끝_번호]에서 시작 번호를 생략하면 문자열의 처음부터 끝 번호까지 뽑아 낸다.

'Life is too short'

- a[시작_번호:끝_번호]에서 시작 번호와 끝 번호를 생략하면 문자열의 처음부터 끝까지 뽑아 낸다.

'Life is too short, You need Python'

- 슬라이싱에서도 인덱싱과 마찬가지로 -(빼기) 기호를 사용할 수 있다.

'You need'

- a[19:-7]은 a[19]에서 a[-8]까지를 의미한다. 이때에도 a[-7]은 포함하지 않는다.

**슬라이싱으로 문자열 나누기**
- 자주 사용하는 슬라이싱기법들

('20230331', 'Rainy')

- 위 예는 문자열 a를 두 부분으로 나누는 기법이다. 
- 숫자 8을 기준으로 문자열 a를 양쪽으로 한 번씩 슬라이싱했다. 
- a[:8]은 a[8]을 포함하지 않고 a[8:]은 a[8]을 포함하기 때문에 8을 기준으로 해서 두 부분으로 나눌 수 있는 것이다. 
- 위 예에서는 "20230331Rainy" 문자열을 날짜를 나타내는 부분인 '20230331'과 날씨를 나타내는 부분인 'Rainy'로 나누는 방법을 보여 준다.

- "20230331Rainy"를 연도인 2023, 월과 일을 나타내는 0331, 날씨를 나타내는 Rainy까지 세 부분으로 나누는 방법은 다음과 같다.

('2023', '0331', 'Rainy')

- 위 예는 숫자 4와 8로 "20230331Rainy" 문자열을 세 부분으로 나누는 방법을 보여 준다.
- 인덱싱과 슬라이싱은 프로그래밍할 때 자주 사용하는 기법이므로 꼭 반복해서 연습해 두자.

**Pithon 문자열을 Python으로 바꾸려면?**

- Pithon 문자열을 Python으로 바꾸려면 어떻게 해야 할까? 제일 먼저 떠오르는 생각은 다음과 같을 것이다.

'i'

TypeError: 'str' object does not support item assignment

- 오류가 나는 이유는 문자열의 요솟값은 바꿀 수 있는 값이 아니기 때문이다
    - 그래서 문자열을 ‘변경 불가능한(immutable) 자료형’이라고도 부른다
- 하지만 앞에서 배운 슬라이싱 기법을 사용하면 Pithon 문자열을 사용해 Python 문자열을 만들 수 있다. 다음 예를 살펴보자.

'P'

'thon'

'Python'

- 슬라이싱을 사용하면 "Pithon" 문자열을 'P' 부분과 'thon' 부분으로 나눌 수 있고, 그 사이에 'y' 문자를 추가하면 'Python'이라는 새로운 문자열을 만들 수 있다.

## 문자열 포매팅이란?

- 문자열에서 또 하나 알아야 할 것으로는 ‘문자열 포매팅(string formatting)’이 있다. 
- 문자열 포매팅을 공부하기 전에 다음과 같은 문자열을 출력하는 프로그램을 작성했다고 가정해 보자.

> "현재 온도는 18도입니다."

- 시간이 지나서 20도가 되면 다음 문장을 출력한다.

> "현재 온도는 20도입니다"

- 두 문자열은 모두 같은데 20이라는 숫자와 18이라는 숫자만 다르다. 
- 이렇게 문자열 안의 특정한 값을 바꿔야 할 경우가 있을 때 이것을 가능하게 해 주는 것이 바로 문자열 포매팅이다.
- 쉽게 말해 문자열 포매팅이란 문자열 안에 어떤 값을 삽입하는 방법이다. 

## 문자열 포매팅 따라 하기

**1. 숫자 바로 대입**

'I eat 3 apples.'

- 결괏값을 보면 알겠지만, 위 예제는 문자열 안에 정수 3을 삽입하는 방법을 보여 준다. 
- 문자열 안의 숫자를 넣고 싶은 자리에 %d 문자를 넣어 주고 삽입할 숫자 3은 가장 뒤에 있는 % 문자 다음에 써 넣었다. 
- 여기에서 %d는 ‘문자열 포맷 코드’라고 부른다.

**2. 문자열 바로 대입**
- 문자열 안에 꼭 숫자만 넣으라는 법은 없다. 이번에는 숫자 대신 문자열을 넣어 보자.

'I eat five apples.'

- 문자열 안에 또 다른 문자열을 삽입하기 위해 앞에서 사용한 문자열 포맷 코드 %d가 아닌 %s를 썼다. 
- 어쩌면 눈치 빠른 학생은 숫자를 넣기 위해서는 %d, 문자열을 넣기 위해서는 %s를 써야 한다는 사실을 눈치챘을 것이다.

> 앞에서 배운 것처럼 문자열을 대입할 때는 반드시 큰따옴표나 작은따옴표를 써야 한다.

**3. 숫자 값을 나타내는 변수로 대입**

'I eat 3 apples.'

- 1번처럼 숫자를 바로 대입하든, 위 예제처럼 숫자 값을 나타내는 변수를 대입하든 결과는 같다.

**4. 2개 이상의 값 넣기**
- 그렇다면 문자열 안에 1개가 아닌 여러 개의 값을 넣고 싶을 때는 어떻게 해야 할까?

'I ate 10 apples. so I was sick for three days.'

- 2개 이상의 값을 넣으려면 마지막 % 다음 괄호 안에 쉼표(,)로 구분하여 각각의 값을 넣어 주면 된다.

## 문자열 포맷 코드
- 문자열 포매팅 예제에서는 대입해 넣는 자료형으로 정수와 문자열을 사용했지만, 이 밖에도 다양한 것을 대입할 수 있다. 
- 문자열 포맷 코드의 종류는 다음과 같다.

| 코드 | 설명                   |
|:-----|:-----------------------|
| %s   | 문자열(String)        |
| %c   | 문자 1개(character)   |
| %d   | 정수(Integer)         |
| %f   | 부동소수(floating-point) |
| %o   | 8진수                 |
| %x   | 16진수                |
| %%   | Literal % (문자 % 자체) |



- 여기에서 재미있는 것은 %s 포맷 코드인데, 이 코드에는 어떤 형태의 값이든 변환해 넣을 수 있다. 무슨 말인지 예를 통해 확인해 보자.

'I have 3 apples'

'rate is 3.234'

- 3을 문자열 안에 삽입하려면 %d를 사용해야 하고 3.234를 삽입하려면 %f를 사용해야 한다. 
- 하지만 %s를 사용하면 %s는 자동으로 % 뒤에 있는 3이나 3.234와 같은 값을 문자열로 바꾸어 대입하기 때문에 이런 것을 생각하지 않아도 된다.

**포매팅 연산자 %d와 %를 같이 쓸 때는 %%를 쓴다**

- "Error is 98%." 가 되게 출력해 보자.

ValueError: incomplete format

- 결괏값으로 당연히 "Error is 98%."가 출력될 것이라고 예상하겠지만, 파이썬은 '형식이 불완전하다'라는 오류 메시지를 보여 준다.
- 그 이유는 ‘문자열 포맷 코드인 %d와 %가 같은 문자열 안에 존재하는 경우, %를 나타내려면 반드시 %%를 써야 한다’라는 법칙이 있기 때문이다. 
    -  %가 포맷 코드와 함께 사용될 때는 포맷 코드와 혼동을 방지하기 위해 %를 나타내려면 %%를 사용해야 
- 이 점은 꼭 기억해 두어야 한다. 하지만 문자열 안에 %d와 같은 포매팅 연산자가 없으면 %는 홀로 쓰여도 상관없다. 
- 따라서 위 예를 제대로 실행하려면 다음과 같이 작성해야 한다.

'Error is 98%.'

## 포맷 코드와 숫자 함께 사용하기

- 앞에서 살펴보았듯이 %d, %s 등과 같은 포맷 코드는 문자열 안에 어떤 값을 삽입할 때 사용한다. 하지만 포맷 코드를 숫자와 함께 사용하면 더 유용하다. 다음 예를 따라해 보자.

**1. 정렬과 공백**

'        hi'

- %10s는 전체 길이가 10개인 문자열 공간에서 대입되는 값을 오른쪽으로 정렬하고 그 앞의 나머지는 공백으로 남겨 두라는 의미이다.

<img src="https://wikidocs.net/images/page/13/02_2_formating1.png">

- 그렇다면 반대쪽인 왼쪽 정렬은 %-10s가 될 것이다.

'hi        jane.'

- hi를 왼쪽으로 정렬하고 나머지는 공백으로 채웠다는 것을 알 수 있다.

<img src="https://wikidocs.net/images/page/13/02_2_formating2.png">

**2. 소수점 표현하기**

'3.4213'

- 3.42134234를 소수점 네 번째 자리까지만 나타내고 싶은 경우에는 위와 같이 작성한다. %0.4f에서 ‘.’는 소수점 포인트, 그 뒤의 숫자 4는 소수점 뒤에 나올 숫자의 개수를 말한다. 소수점 포인트 앞의 숫자는 문자열의 전체 길이를 의미하는데, %0.4f에서 사용한 숫자 0은 길이에 상관하지 않겠다는 의미이다.

> %0.4f는 0을 생략하여 %.4f처럼 사용하기도 한다.

'    3.4213'

- 위는 숫자 3.42134234를 소수점 네 번째 자리까지만 표시하고 전체 길이가 10개인 문자열 공간에서 오른쪽으로 정렬하는 예를 보여 준다.

<img src="https://wikidocs.net/images/page/13/02_2_formating3.png">

## format 함수를 사용한 포매팅

- 문자열의 format 함수를 사용하면 좀 더 발전된 스타일로 문자열 포맷을 지정할 수 있다. 앞에서 살펴본 문자열 포매팅 예제를 format 함수를 사용해서 바꾸면 다음과 같다.

**숫자 바로 대입하기**

'I eat 3 apples'

**문자열 바로 대입하기**

'I eat five apples'

**숫자 값을 가진 변수로 대입하기**

'I eat 3 apples'

**2개 이상의 값 넣기**

'I ate 10 apples. so I was sick for three days.'

- 2개 이상의 값을 넣을 경우, 문자열의 {0}, {1}과 같은 인덱스 항목이 format 함수의 입력값으로 순서에 맞게 바뀐다. 
- 위 예에서 {0}은 format 함수의 첫 번째 입력값인 number, {1}은 format 함수의 두 번째 입력값인 day로 바뀐다.

**이름으로 넣기**

'I ate 10 apples. so I was sick for 3 days.'

- {0}, {1}과 같은 인덱스 항목 대신 더 편리한 {name} 형태를 사용하는 방법도 있다. {name} 형태를 사용할 경우, format 함수에는 반드시 name=value와 같은 형태의 입력값이 있어야 한다. 위 예는 문자열의 {number}, {day}가 format 함수의 입력값인 number=10, day=3 값으로 각각 바뀌는 것을 보여 주고 있다.

**인덱스와 이름을 혼용해서 넣기**

'I ate 10 apples. so I was sick for 3 days.'

- 인덱스 항목과 name=value 형태를 혼용하는 것도 가능하다.

**왼쪽 정렬**

'hi        '

- :<10 표현식을 사용하면 치환되는 문자열을 왼쪽으로 정렬하고 문자열의 총 자릿수를 10으로 맞출 수 있다.

**오른쪽 정렬**

'        hi'

- 오른쪽 정렬은 :< 대신 :>을 사용하면 된다. 화살표의 방향을 생각하면 어느 쪽으로 정렬되는지 바로 알 수 있을 것이다.

**가운데 정렬**

'    hi    '

- :^를 사용하면 가운데 정렬도 가능하다.

**공백 채우기**

'====hi===='

'hi!!!!!!!!'

- 정렬할 때 공백 문자 대신 지정한 문자 값으로 채워 넣을 수도 있다. 채워 넣을 문자 값은 정렬 문자 <, >, ^ 바로 앞에 넣어야 한다. 위 예에서 첫 번째 예제는 가운데(^)로 정렬하고 빈 공간을 =로 채웠고, 두 번째 예제는 왼쪽(<)으로 정렬하고 빈 공간을 !로 채웠다.

**소수점 표현하기**

'3.4213'

- 위는 format 함수를 사용해 소수점을 4자리까지만 표현하는 방법을 보여 준다. 앞에서 살펴보았던 표현식 0.4f를 그대로 사용한 것을 알 수 있다.

'    3.4213'

- 위와 같이 자릿수를 10으로 맞출 수도 있다. 이 또한 앞에서 살펴본 10.4f의 표현식을 그대로 사용한 것을 알 수 있다.

**{ 또는 } 문자 표현하기**

'{ and }'

- format 함수를 사용해 문자열을 포매팅할 경우, {}와 같은 중괄호 문자를 포매팅 문자가 아닌 문자 그대로 사용하고 싶은 경우에는 위 예의 {{}}처럼 2개를 연속해서 사용하면 된다.

## f 문자열 포매팅

- 파이썬 3.6 버전부터는 f 문자열 포매팅 기능을 사용할 수 있다. 파이썬 3.6 미만 버전에서는 사용할 수 없는 기능이므로 주의해야 한다.
- 다음과 같이 문자열 앞에 f 접두사를 붙이면 f 문자열 포매팅 기능을 사용할 수 있다.

'나의 이름은 홍길동입니다. 나이는 30입니다.'

- f 문자열 포매팅은 위와 같이 name, age와 같은 변숫값을 생성한 후에 그 값을 참조할 수 있다. 또한 f 문자열 포매팅은 표현식을 지원하기 때문에 다음과 같은 것도 가능하다.

> 표현식이란 중괄호 안의 변수를 계산식과 함께 사용하는 것을 말한다.

'나는 내년이면 31살이 된다.'

- 딕셔너리는 f 문자열 포매팅에서 다음과 같이 사용할 수 있다.

'나의 이름은 홍길동입니다. 나이는 30입니다.'

> 딕셔너리는 Key와 Value라는 것을 한 쌍으로 가지는 자료형이다. 02-5에서 자세히 알아본다.

- 정렬은 다음과 같이 할 수 있다.

In [108]:
  # 왼쪽 정렬

'hi        '

In [109]:
  # 오른쪽 정렬

'        hi'

In [110]:
  # 가운데 정렬

'    hi    '

- 공백 채우기는 다음과 같이 할 수 있다.

In [111]:
  # 가운데 정렬하고 '=' 문자로 공백 채우기

'====hi===='

In [112]:
  # 왼쪽 정렬하고 '!' 문자로 공백 채우기

'hi!!!!!!!!'

- 소수점은 다음과 같이 표현할 수 있다.

In [114]:
y = 3.42134234
  # 소수점 4자리까지만 표현

'3.4213'

In [115]:
  # 소수점 4자리까지 표현하고 총 자리수를 10으로 맞춤

'    3.4213'

- f 문자열에서 {}를 문자 그대로 표시하려면 다음과 같이 2개를 동시에 사용해야 한다.

'{ and }'

**문제**
- format 함수 또는 f 문자열 포메팅을 사용, 다음 문자열을 출력해 보자

> !!!python!!!

'!!!python!!!'

## 문자열 관련 함수들
- 문자열 자료형은 자체적으로 함수를 가지고 있다. 이들 함수를 다른 말로 ‘문자열 내장 함수’라고 한다. 이 내장 함수를 사용하려면 문자열 변수 이름 뒤에 ‘.’를 붙인 후 함수 이름을 써 주면 된다. 이제 문자열의 내장 함수에 대해서 알아보자.

**문자 개수 세기 - count**

In [120]:
a = "hobby"


2

- count 함수로 문자열 중 문자 b의 개수를 리턴했다.

**위치 알려 주기 1 - find**

In [121]:
a = "Python is the best choice"


14

-1

- find 함수로 문자열 중 문자 b가 처음으로 나온 위치를 반환했다. 만약 찾는 문자나 문자열이 존재하지 않는다면 -1을 반환한다.

> 파이썬은 숫자를 0부터 세기 때문에 b의 위치는 15가 아닌 14가 된다.

**위치 알려 주기 2 - index**

In [123]:
a = "Life is too short"


8

8

ValueError: substring not found

-1

- index 함수로 문자열 중 문자 t가 맨 처음으로 나온 위치를 반환했다. 만약 찾는 문자나 문자열이 존재하지 않는다면 오류가 발생한다. 앞의 find 함수와 다른 점은 문자열 안에 존재하지 않는 문자를 찾으면 오류가 발생한다는 것이다.

**문자열 삽입 - join**

'a,b,c,d'

- join 함수로 abcd 문자열의 각각의 문자 사이에 ‘,’를 삽입했다.

- join 함수는 문자열뿐만 아니라 앞으로 배울 리스트나 튜플도 입력으로 사용할 수 있다(리스트와 튜플은 곧 배울 내용이므로 여기에서는 잠시 눈으로만 살펴보자). join 함수의 입력으로 리스트를 사용하는 예는 다음과 같다.

'a,b,c,d'

**소문자를 대문자로 바꾸기 - upper**

In [130]:
a = "hi"


'HI'

- upper 함수는 소문자를 대문자로 바꾸어 준다. 만약 문자열이 이미 대문자라면 아무런 변화도 일어나지 않을 것이다.

**대문자를 소문자로 바꾸기 - lower**

In [132]:
a = "HI"


'hi'

- lower 함수는 대문자를 소문자로 바꾸어 준다.

**왼쪽 공백 지우기 - lstrip**

In [133]:
a = " hi "


'hi '

- lstrip 함수는 문자열 중 가장 왼쪽에 있는 한 칸 이상의 연속된 공백들을 모두 지운다. lstrip에서 l은 left를 의미한다.

**오른쪽 공백 지우기 - rstrip**

In [134]:
a= " hi "


' hi'

- rstrip 함수는 문자열 중 가장 오른쪽에 있는 한 칸 이상의 연속된 공백을 모두 지운다. rstrip에서 r은 right를 의미한다.

**양쪽 공백 지우기 - strip**

In [135]:
a = " hi "


'hi'

- strip 함수는 문자열 양쪽에 있는 한 칸 이상의 연속된 공백을 모두 지운다.

**문자열 바꾸기 - replace**

In [136]:
a = "Life is too short"


'Your leg is too short'

- replace 함수는 replace(바뀔_문자열, 바꿀_문자열)처럼 사용해서 문자열 안의 특정한 값을 다른 값으로 치환해 준다.

**문자열 나누기 - split**

In [137]:
a = "Life is too short"


['Life', 'is', 'too', 'short']

In [138]:
b = "a:b:c:d"


['a', 'b', 'c', 'd']

- split 함수는 a.split()처럼 괄호 안에 아무 값도 넣어 주지 않으면 공백([Space]], [Tab], [Enter])을 기준으로 문자열을 나누어 준다. 만약 b.split(':')처럼 괄호 안에 특정 값이 있을 경우에는 괄호 안의 값을 구분자로 해서 문자열을 나누어 준다. 이렇게 나눈 값은 리스트에 하나씩 들어간다. ['Life', 'is', 'too', 'short']나 ['a', 'b', 'c', 'd']가 리스트인데, 02-3에서 자세히 알아볼 것이므로 여기에서는 너무 신경 쓰지 않아도 된다.

**착각하기 쉬운 문자열 함수**
- 소문자를 대문자로 바꾸어 주는 다음의 예를 보자.

In [139]:
a = 'hi'


'HI'

- 이와 같이 실행한 후에 a 변수의 값은 'HI'로 변했을까? 아니면 'hi' 값을 유지할까? 다음과 같이 확인해 보자.

'hi'

- a.upper()를 수행하더라도 a 변수의 값은 변하지 않았다. 왜냐하면 a.upper()를 실행하면 upper 함수는 a 변수의 값 자체를 변경하는 것이 아니라 대문자로 바꾼 값을 리턴하기 때문이다. 문자열은 이전에도 잠깐 언급했지만 자체의 값을 변경할 수 없는 immutable 자료형이다. 따라서 a 값을 'HI' 로 바꾸고 싶다면 다음과 같이 대입문을 사용해야 한다.

'HI'

- upper 뿐만 아니라 lower, join, lstrip, rstrip, strip, replace, split 함수는 모두 이와 같은 규칙이 적용되어 문자열 자체의 값이 변경되는 것이 아니라 변경된 값을 리턴한다는 사실에 주의하자.

**문제**

- 문자열 "hello world"를 대문자로 변환하여 출력하세요.

- 문자열 "Hello World"를 소문자로 변환하여 출력하세요.

- 문자열 "apple,banana,grape"을 쉼표(,)를 기준으로 분리하여 리스트로 출력하세요.

- 문자열 " Python "의 양쪽 공백을 제거하여 출력하세요.

- 문자열 "Python "의 오른쪽 공백을 제거하여 출력하세요.

- 문자열 "I like apples"에서 "apples"를 "bananas"로 대체하여 출력하세요.

# 리스트 자료형

- [소스](https://wikidocs.net/14)
- 지금까지 우리는 숫자와 문자열에 대해서 알아보았다. 하지만 숫자와 문자열만으로 프로그래밍을 하기에는 부족한 점이 많다. 예를 들어 1부터 10까지의 숫자 중 홀수의 모음인 1, 3, 5, 7, 9의 집합을 생각해 보자. 이런 숫자의 모음을 숫자나 문자열로 표현하기는 어렵다. 파이썬에는 이러한 불편함을 해소할 수 있는 자료형이 존재한다. 이것이 바로 지금부터 공부할 리스트(list)이다.

## 리스트는 어떻게 만들고 사용할까?

- 리스트를 사용하면 1, 3, 5, 7, 9의 숫자 모음을 다음과 같이 간단하게 표현할 수 있다.

[1, 3, 5, 7, 9]

- 리스트를 만들 때는 위에서 보는 것과 같이 대괄호([])로 감싸 주고 각 요솟값은 쉼표(,)로 구분해 준다.

> 리스트명 = [요소1, 요소2, 요소3, ...]

- 여러 가지 리스트의 생김새는 다음과 같다.

- 비어 있는 리스트는 a = list()로 생성할 수 있다.

- 리스트는 a처럼 아무것도 포함하지 않아 비어 있는 리스트([])일 수도 있고, b처럼 숫자를 요솟값으로 가질 수도 있으며, c처럼 문자열을 요솟값으로 가질 수도 있다. 또한 d처럼 숫자와 문자열을 함께 요솟값으로 가질 수도 있고, e처럼 리스트 자체를 요솟값으로 가질 수도 있다. 즉, 리스트 안에는 어떠한 자료형도 포함할 수 있다.

<img src='https://wikidocs.net/images/page/14/02_3_list.png'>

## 리스트의 인덱싱과 슬라이싱

- 리스트도 문자열처럼 인덱싱과 슬라이싱이 가능하다. 말로 설명하는 것보다 직접 예를 실행해 보면서 리스트의 기본 구조를 이해하는 것이 쉽다. 대화형 인터프리터로 따라 하면서 확실하게 이해하자.


**리스트의 인덱싱**
- 리스트 역시 문자열처럼 인덱싱을 적용할 수 있다. 먼저 a 변수에 [1, 2, 3] 값을 설정한다.



[1, 2, 3]

- a[0]은 리스트 a의 첫 번째 요솟값을 말한다.


1

- 다음 예는 리스트의 첫 번째 요소인 a[0]과 세 번째 요소인 a[2]의 값을 더한 것이다.



4

- 이것은 1 + 3으로 해석되어 값 4를 출력한다.

- 문자열을 공부할 때 이미 살펴보았지만, 파이썬은 숫자를 0부터 세기 때문에 a[1]이 리스트 a의 첫 번째 요소가 아니라 a[0]이 리스트 a의 첫 번째 요소라는 것을 명심하자. 
- a[-1]은 문자열에서와 마찬가지로 리스트 a의 마지막 요솟값을 말한다.



3

- 이번에는 다음 예처럼 리스트 a를 숫자 1, 2, 3과 또 다른 리스트인 ['a', 'b', 'c']를 포함하도록 만들어 보자.



[1, 2, 3, ['a', 'b', 'c']]

1

['a', 'b', 'c']

['a', 'b', 'c']

- 예상한 대로 a[-1]은 마지막 요솟값 ['a', 'b', 'c']를 나타낸다. a[3]은 리스트 a의 네 번째 요소를 나타내기 때문에 마지막 요소를 나타내는 a[-1]과 동일한 결괏값을 보여 준다.

- 그렇다면 리스트 a에 포함된 ['a', 'b', 'c'] 리스트에서 'a' 값을 인덱싱을 사용해 끄집어 낼 수 있는 방법은 없을까? 다음 예를 살펴보자.

'a'

- 위와 같이 하면 'a'를 끄집어 낼 수 있다. a[-1]이 ['a', 'b', 'c'] 리스트라는 것은 이미 설명했다. 바로 이 리스트에서 첫 번째 요소를 불러오기 위해 [0]을 붙여 준 것이다.

- 다음 예도 마찬가지 경우이므로 어렵지 않게 이해할 수 있을 것이다.



'b'

'c'

**삼중 리스트에서 인덱싱하기**
- 조금 복잡하지만, 리스트를 다음과 같이 작성할 수도 있다.



In [156]:
a = [1, 2, ['a', 'b', ['Life', 'is']]]

- 리스트 a 안에 ['a', 'b', ['Life', 'is']] 리스트가 포함되어 있고 그 리스트 안에 다시 ['Life', 'is'] 리스트가 포함되어 있다. 즉, 삼중 구조의 리스트인 것이다.

- 이 경우, 'Life' 문자열만 끄집어 내려면 다음과 같이 해야 한다.


'Life'

- 위 예는 리스트 a의 세 번째 요소인 리스트 ['a', 'b', ['Life', 'is']]에서 세 번째 요소인 리스트 ['Life', 'is']의 첫 번째 요소를 나타낸다.

- 리스트를 다중으로 중첩해서 사용하는 것은 혼란스럽기 때문에 자주 사용하지는 않는다.

**리스트의 슬라이싱**
- 문자열과 마찬가지로 리스트에서도 슬라이싱 기법을 적용할 수 있다. 슬라이싱은 ‘잘라 낸다’라는 뜻이라고 했다. 리스트의 슬라이싱에 대해서 살펴보자.



[1, 2]

- 앞의 예를 문자열에서 슬라이싱했던 예와 비교해 보자.



'12'

- 2가지가 완전히 동일하게 사용되었다는 것을 눈치챘을 것이다. 문자열에서 했던 것과 사용법이 완전히 동일하다.

- 몇 가지 예를 더 살펴보자.

[1, 2]

[3, 4, 5]

- b 변수는 리스트 a의 첫 번째 요소부터 두 번째 요소인 a[1]까지 나타내는 리스트이다. 물론 a[2] 값인 3은 포함되지 않는다. c라는 변수는 리스트 a의 세 번째 요소부터 끝까지 나타내는 리스트이다.




**문제**
- a = [1,2 3, 4, 5] 리스트에서, 슬라이싱 기법을 사용하여 리스트 [2, 3] 을 만들어 보자

**중첩된 리스트에서 슬라이싱하기**
- 리스트가 포함된 중첩 리스트 역시 슬라이싱 방법은 똑같이 적용된다.



[3, ['a', 'b', 'c'], 4]

['a', 'b']

- 위 예에서 a[3]은 ['a', 'b', 'c']를 나타낸다. 따라서 a[3][:2]는 ['a', 'b', 'c']의 첫 번째 요소부터 세 번째 요소 직전까지의 값, 즉 ['a', 'b']를 나타내는 리스트가 된다.


## 리스트 연산하기
- 리스트 역시 +를 사용해서 더할 수 있고 *를 사용해서 반복할 수 있다. 문자열과 마찬가지로 리스트에서도 되는지 직접 확인해 보자.


**리스트 더하기(+)**

[1, 2, 3, 4, 5, 6]

- 리스트 사이에서 +는 2개의 리스트를 합치는 기능을 한다. 문자열에서 "abc" + "def" = "abcdef"가 되는 것과 같은 이치이다.



**리스트 반복하기(*)**

[1, 2, 3, 1, 2, 3, 1, 2, 3]

- 위에서 볼 수 있듯이 [1, 2, 3] 리스트가 세 번 반복되어 새로운 리스트를 만들어 낸다. 문자열에서 "abc" * 3 = "abcabcabc" 가 되는 것과 같은 이치이다.



**리스트 길이 구하기**
- 리스트 길이를 구하기 위해서는 다음처럼 len 함수를 사용해야 한다.



3

- len은 문자열, 리스트 외에 앞으로 배울 튜플과 딕셔너리에도 사용할 수 있는 함수이다. 실습에서 자주 사용하므로 잘 기억해 두자.




**초보자가 범하기 쉬운 리스트 연산 오류**
- 다음 소스 코드를 입력했을 때 결괏값은 어떻게 나올까?



- a[2]의 값인 3과 문자열 hi가 더해져서 3hi가 출력될 것이라고 생각할 수 있다. 

- 하지만 다음 결과를 살펴보니 오류가 발생했다. 오류의 원인은 무엇일까?
- a[2]에 저장된 값은 3이라는 정수인데 "hi"는 문자열이다. 정수와 문자열은 당연히 서로 더할 수 없기 때문에 오류가 발생한 것이다.

- 만약 숫자와 문자열을 더해서 '3hi'를 만들고 싶다면 다음처럼 숫자 3을 문자 '3'으로 바꾸어야 한다.

'3hi'

## 리스트의 수정과 삭제
- 리스트는 값을 수정하거나 삭제할 수 있다.



**리스트의 값 수정하기**


[1, 2, 4]

- a[2]의 요솟값 3이 4로 바뀌었다.


**del 함수를 사용해 리스트 요소 삭제하기**

[1, 3]

- del a[x]는 x번째 요솟값을 삭제한다. 위에서는 a 리스트에서 a[1]을 삭제했다. 
- del 함수는 파이썬이 자체적으로 가지고 있는 삭제 함수이며 다음과 같이 사용한다.

> del 객체

- 객체란 파이썬에서 사용되는 모든 자료형을 말한다.
- 다음처럼 슬라이싱 기법을 사용하여 리스트의 요소 여러 개를 한꺼번에 삭제할 수도 있다.


[1, 2]

- a[2:]에 해당하는 리스트의 요소들이 삭제되었다.

- 리스트의 요소를 삭제하는 방법에는 2가지가 더 있다. 바로 리스트의 remove와 pop 함수를 사용하는 것인데, 이는 리스트 관련 함수에서 설명한다.

## 리스트 관련 함수
- 문자열과 마찬가지로 리스트 변수 이름 뒤에 ‘.’를 붙여 여러 가지 리스트 관련 함수를 사용할 수 있다. 유용하게 사용하는 리스트 관련 함수 몇 가지만 알아보자.

**리스트에 요소 추가하기 - append**
- append의 사전적 의미는 ‘덧붙이다, 첨부하다’이다. 이 뜻을 안다면 다음 예가 바로 이해될 것이다. append(x)는 리스트의 맨 마지막에 x를 추가하는 함수이다.


[1, 2, 3, 4]

- 리스트 안에는 어떤 자료형도 추가할 수 있다. 다음 예는 리스트에 다시 리스트를 추가한 결과이다.



[1, 2, 3, 4, [5, 6]]

**리스트 정렬 - sort**
- sort 함수는 리스트의 요소를 순서대로 정렬해 준다.



[1, 2, 3, 4]

**리스트 뒤집기 - reverse**
- reverse 함수는 리스트를 역순으로 뒤집어 준다. 이때 리스트 요소들을 순서대로 정렬한 다음 다시 역순으로 정렬하는 것이 아니라 현재의 리스트를 그대로 거꾸로 뒤집는다.



['b', 'c', 'a']

**인덱스 반환 - index**
- index(x) 함수는 리스트에 x 값이 있으면 x의 인덱스 값(위칫값)을 리턴한다.


2

0

- 위 예에서 리스트 a에 있는 숫자 3의 위치는 a[2]이므로 2, 숫자 1의 위치는 a[0]이므로 0을 리턴한다.

- 다음 예에서 값 0은 a 리스트에 존재하지 않기 때문에 오류가 발생한다.


ValueError: 0 is not in list

**리스트에 요소 삽입 - insert**
- insert(a, b)는 리스트의 a번째 위치에 b를 삽입하는 함수이다. 파이썬은 숫자를 0부터 센다는 것을 반드시 기억하자.



[4, 1, 2, 3]

- 위 예는 0번째 자리, 즉 첫 번째 요소인 a[0] 위치에 값 4를 삽입하라는 뜻이다.


[4, 1, 2, 5, 3]

- 위 예는 리스트 a의 a[3], 즉 네 번째 요소 위치에 값 5를 삽입하라는 뜻이다.



**리스트 요소 제거 - remove**
- remove(x)는 리스트에서 첫 번째로 나오는 x를 삭제하는 함수이다.



[1, 2, 1, 2, 3]

- a가 3이라는 값을 2개 가지고 있을 경우, 첫 번째 3만 제거되는 것을 알 수 있다.



[1, 2, 1, 2]

**리스트 요소 끄집어 내기 - pop**
- pop()은 리스트의 맨 마지막 요소를 리턴하고 그 요소는 삭제한다.



[1, 2]

- a.pop(1)은 a[1]의 값을 끄집어 내어 리턴한다. 다시 a를 출력해 보면 끄집어 낸 값이 삭제된 것을 확인할 수 있다.



**리스트에 포함된 요소 x의 개수 세기 - count**
- count(x)는 리스트 안에 x가 몇 개 있는지 조사하여 그 개수를 리턴하는 함수이다.



2

- 1이라는 값이 리스트 a에 2개 들어 있으므로 2를 리턴한다.



**리스트 확장 - extend**
- extend(x)에서 x에는 리스트만 올 수 있으며 원래의 a 리스트에 x 리스트를 더하게 된다.



[1, 2, 3, 4, 5]

[1, 2, 3, 4, 5, 6, 7, 6, 7]

- a.extend([4, 5])는 a += [4, 5]와 동일하다.

> a += [4, 5]는 a = a + [4, 5]와 동일한 표현식이다.



**문제**

- 리스트 ["apple", "banana", "grape"]를 생성하고, 세 번째 요소를 인덱싱하여 출력하세요.

- 리스트 [1, 2, 3, 4, 5]를 생성하고, 두 번째부터 네 번째까지의 요소를 슬라이싱하여 출력하세요.

- 리스트 ["apple", "banana", "grape"]에 "orange"를 추가하세요.

- 리스트 [1, 2, 3, 4, 5]에서 세 번째 요소를 10으로 수정하세요.

- 리스트 [1, 2, 3, 4, 5]에서 네 번째 요소를 삭제하세요.

- 정수로 이루어진 리스트 [3, 1, 4, 2, 5]를 생성하고, 이를 오름차순으로 정렬한 결과를 출력하세요.

- 문자열로 이루어진 리스트 ["apple", "banana", "grape"]를 생성하고, 이를 역순으로 뒤집은 결과를 출력하세요.

- 정수로 이루어진 리스트 [1, 2, 3, 4, 5]에서 값이 3인 요소의 인덱스를 찾아 출력하세요.


- 정수로 이루어진 리스트 [1, 2, 3, 5]에 값이 4인 요소를 숫자의 순서에 맞게 삽입하여 결과를 출력하세요.


- 정수로 이루어진 리스트 [1, 2, 3, 4, 5]에서 값이 3인 요소를 제거하여 결과를 출력하세요.


- 정수로 이루어진 리스트 [1, 2, 3, 4, 5]에서 세 번째 요소를 제거하고 반환한 후, 결과를 출력하세요.


- 정수로 이루어진 리스트 [1, 2, 2, 3, 2, 4, 2, 5]에서 값이 2인 요소의 개수를 계산하여 출력하세요.


- 리스트 [1, 2, 3]과 [4, 5, 6]을 합쳐서 출력하세요.

# 튜플 자료형

# 딕셔너리 자료형

# 집합 자료형

# 불 자료형

# 자료형의 값을 저장하는 공간, 변수